### ISSUE TITLE
Non-determinism in some Predictors with dropout

### ISSUE THREAD

**Issue by jacobdanovitch** [allenai/allennlp#3888]
**Describe the bug**
When loading a Predictor directly from its constructor, the underlying model is kept in training mode. This causes models with dropout to apply it during inference, which of course leads to non-deterministic predictions.

**Steps to reproduce**
Creating some custom Predictor and creating it directly from the constructor (`predictor = MyPredictor(model, dataset_reader)`) will yield True for `predictor._model.training`. This does _not_ apply when loading from_path or from_archive etc.

Not sure if this is something the team feels would need to be fixed or is on the user to be aware of (or if it was documented somewhere and I've missed it). For anyone experiencing the same issue, this can be fixed by using `super().__init__(model.eval(), dataset_reader)` in your Predictor's init.

**Comment by matt-gardner** [comment]
Thanks for catching this!  I think it's definitely something we'd be happy to have a PR to clarify, either by updating a docstring or by adding a parameter to do this.  I'd probably be ok with a parameter which, if true, would call `model.eval()` before passing it to `super().__init__()`, and then the docstring for that parameter could describe the issue.

**Comment by jacobdanovitch** [comment]
> Thanks for catching this! I think it's definitely something we'd be happy to have a PR to clarify, either by updating a docstring or by adding a parameter to do this. I'd probably be ok with a parameter which, if true, would call `model.eval()` before passing it to `super().__init__()`, and then the docstring for that parameter could describe the issue.

Sounds good! I'll get a PR up in the next day or two.
